File: sdk-python/src/personaai/chat/_sse.py
"""Pure SSE-framing/event-decoding logic, shared by the sync and async chat
clients (each drives it from its own kind of chunk iterator, so the actual
`for`/`async for` loop can't be shared, only this buffer/decode step can).
Ported from ``sdk/src/chat/sse.ts``'s ``parseAguiEventStream``."""

from __future__ import annotations

import json as json_lib

from ..types.chat import AguiEvent


def split_buffer(buffer: str) -> tuple[list[str], str]:
    """Splits ``buffer`` on the SSE frame delimiter (``\\n\\n``), returning
    (complete raw frames, leftover partial frame) — mirrors
    ``buffer.split('\\n\\n')`` + ``.pop()`` in ``sse.ts``."""
    frames = buffer.split("\n\n")
    leftover = frames.pop()
    return frames, leftover


def decode_frame(raw_frame: str) -> AguiEvent | None:
    """Extracts and JSON-decodes the ``data:`` line(s) of one SSE frame.
    Returns ``None`` for frames with no data, or malformed/partial JSON —
    skipped rather than raised, matching SSE's own tolerant-parsing
    convention."""
    data_lines = [line[5:].lstrip() for line in raw_frame.split("\n") if line.startswith("data:")]
    if not data_lines:
        return None

    data = "\n".join(data_lines)
    if not data:
        return None

    try:
        event = json_lib.loads(data)
    except ValueError:
        return None

    return event if isinstance(event, dict) else None

File: sdk/src/chat/sse.ts
import type { AguiEvent } from '../types/chat.js';

/**
 * Parses a `text/event-stream` `Response` body into a stream of AG-UI
 * events. Hand-rolled rather than using `@ag-ui/client`'s `HttpAgent` —
 * that package pulls in rxjs/fast-json-patch/uuid (a much heavier
 * dependency footprint than this SDK's "native fetch, no heavy deps"
 * design established since `HttpClient`), and its `RunAgentInput`-body
 * threadId semantics don't match this backend's own header-based routing
 * convention (`x-agent-id`/`x-thread-id`, confirmed by reading
 * `developerAgui.routes.js` directly) — spiked and rejected, not skipped.
 * `@ag-ui/core` alone (zod-only, no rxjs) supplies accurate event typing.
 * @param response - A `Response` whose body is a `text/event-stream`.
 * @yields Each parsed {@link AguiEvent}; malformed/partial SSE frames are
 *   skipped rather than thrown.
 */
export async function* parseAguiEventStream(response: Response): AsyncGenerator<AguiEvent> {
  if (!response.body) return;

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const events = buffer.split('\n\n');
      buffer = events.pop() ?? '';

      for (const rawEvent of events) {
        const dataLines = rawEvent
          .split('\n')
          .filter((line) => line.startsWith('data:'))
          .map((line) => line.slice(5).trimStart());
        if (dataLines.length === 0) continue;

        const data = dataLines.join('\n');
        if (!data) continue;

        try {
          yield JSON.parse(data) as AguiEvent;
        } catch {
          // Malformed/partial SSE frame — skip rather than throw, matching
          // SSE's own tolerant-parsing convention.
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
}
